### DISCUSSION TITLE
Close : Report on display speed.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#171] [General]
I start to use BiblioteQ in real condition : report on display speed

configuration :
Show Progress Dialogs disabled
Show Table / View Image disabled
Entrie per Page / Unlimited Entries per Page

Result:
by menu View/Category/Books : about **12 seconds** to display my 7106 documents on 24 columns (with availability and totalreserved)

by sql about **3 seconds** : "SELECT book_read,title,author,publisher,pdate,place,edition,category,language,id,price,monetary_units,quantity,binding_type,location,isbn13,lccontrolnumber,callnumber,deweynumber,originality,condition,accession_number FROM book"
**availability,totalreserved" is missing**, I didn't find it in the sql tables to have the 24 columns of the Books view

by sql about **6 seconds** for "SELECT * FROM book" on 30 columns

I do not understand that we arrive at a factor of 4 between display by sql and the menu to display 2 more fields
can you look at this point, thank you for the time spent at BiblioteQ

**Comment by textbrowser** [comment]
Can you upload the database here? There are several explanations. First, BQ has to create 7106 * N display items. It then has to populate each of those items. Second, the Books query includes sub-queries for computing sums whereas your queries do not.

**Comment by textbrowser** [comment]
Also, Qt drivers do not necessarily support the size of a query's result set. I believe for PQ that they do and not for SQLite. As a result, each row of the table has to be added for a particular record. Setting the row count once is better. Some of the queries in BQ have corresponding count queries; this varies with the complexity of a query.

**Comment by meteos77** [comment]
if all is normal it's great it was just an observation 
my test base
[2022-06-Bibliothèque.sqlite.zip](https://github.com/textbrowser/biblioteq/files/7829660/2022-06-Bibliotheque.sqlite.zip)

**Comment by textbrowser** [comment]
I see. Yes, it lacks performance.

**Reply by meteos77** [reply]
my favorite sql query :-) no lack of performance
select accession_number,title,author,callnumber,isbn13,publisher from book

**Reply by meteos77** [reply]
small correction if you want to edit after :
select accession_number,title,author,callnumber,isbn13,publisher,book.myoid, book.type from book

**Comment by textbrowser** [comment]
BQ spends 86 milliseconds on the query. As suspected, it's the 7106 * N items.

**Comment by textbrowser** [comment]
BQ also has to create the other view because the person way wish to see a graphical view of the contents. See View -> Icons Mode.

**Comment by textbrowser** [comment]
So it is creating two views.

**Reply by meteos77** [reply]
even if you never use it?

**Comment by textbrowser** [comment]
Sure. If you change views, it would have to perform the query again. It does not have to if it caches the entire result. It has the cache in the table view but not in the icon view.

**Reply by meteos77** [reply]
but I have no image in the icon view, it takes time to display nothing?

**Comment by textbrowser** [comment]
It creates default images.

**Comment by textbrowser** [comment]
BQ prefers pagination. The unlimited size was again a request. Users.

**Comment by textbrowser** [comment]
https://textbrowser.github.io/biblioteq/

May 15, 2011.

**Comment by meteos77** [comment]
we all have our own way of working and the table view for me allows to make a classification and to detect the errors and correct them.

practical exercise
there are 7 errors on the name of an author
Nothomb Am#lie instead of Nothom Amélie.

I put the table in unlimited, I classify by author and I see the 7 documents that I must modify.
with the pagination mode how do you do it ?

**Comment by meteos77** [comment]
this mode is made for the detection of the typing errors (with the resumption of the existing and it is very well like that)
after in exploitation my screen displays 30 lines, that of the library in 20 and 23 thus mode pagination.

it will change the habits :-) but the people use the table view only when it meets a problem of shower / accession_number not valid. With BiblioteQ, there will be no more problem so the table view will be a bonus :-)

**Comment by textbrowser** [comment]
9176 ms for populating the table.

**Reply by meteos77** [reply]
what does population the table represent

**Comment by textbrowser** [comment]
8328 without tool tips.

**Comment by textbrowser** [comment]
7597 without special ISBN formatting.

**Comment by meteos77** [comment]
it's great it gives a perf table according to the requested options 
and without the image function ?

**Comment by textbrowser** [comment]
Between 8 and 13 seconds.

**Comment by textbrowser** [comment]
7.5 (icons and table) without book-read status.

**Reply by meteos77** [reply]
the without are cumulative ?

is it without tool tips + without isbn + without book-read or is it individually?

**Comment by textbrowser** [comment]
Cannot disable book-read status.

**Comment by textbrowser** [comment]
Another user request; the book-read status.

**Reply by meteos77** [reply]
it is the only field that can be modified in the table it is privileged :-)

**Comment by textbrowser** [comment]
I'll add a new option to disable the icons if desired; with a nice warning.

**Reply by meteos77** [reply]
a warning for what purpose?

**Reply by textbrowser** [reply]
That the Icons View will be empty.

**Reply by meteos77** [reply]
I don't see how we work with the icon view and 100 visible documents.
can you explain it briefly please ?

**Reply by textbrowser** [reply]
The Icons View displays icons of items in a grid; without a table.

**Reply by meteos77** [reply]
I looked at my 47 book base with the icons
I can't use it because it must represent a huge size of base to store all the covers and the pc and the backup won't support it. but it gives me some ideas...

on the other hand it is beautiful, I would compare it to a museum, we look at the works but we do not have the labels (a small tool tip in the icon view (if the 2 are generated)? at the 1st click tools tip and at the 2nd the editor

**Reply by meteos77** [reply]
ask really if it's a 5 minutes affair to realize if not a very distant to do list

idea : export of the icon view by page of 25 in pdf format

my idea is to allow members to receive a newletter with the new releases for example "Les Romans Policiers" and to make a pdf with the cover images.

![exemple](https://user-images.githubusercontent.com/56431039/148584899-31505118-ab70-4593-94f9-d82287cc6df8.png)
:

**Reply by textbrowser** [reply]
Completed as a print option. See File. I do not recommend exporting all of the content. It works but it's very small. 100 items per view works well. :)

**Comment by textbrowser** [comment]
100 items in about 212 ms.

**Reply by meteos77** [reply]
I redid the test at the beginning with the options 
show progress dialog enabled
show progress Table View
show tool tips
 I was at 58 seconds on my machine for 7106

**Comment by textbrowser** [comment]
Lots of new optimizations! :) New option for disabling the Read Status column along with a tool tip. See Other Options.

**Reply by meteos77** [reply]
Stunning! a big THANK YOU for these great optimizations  < 4 seconds instead of 12 for a factor of 3 less ! 

Result with the same tests as in the beginning

by menu View/Category/Books : about  **<8** (instead of 12 seconds) to display my 7106 documents on 24 columns (with availability and totalreserved)

by menu View/Category/Books without read-status : about **<4** (instead of 12 seconds) to display my 7106 documents on 24 columns (with availability and totalreserved)

by sql about **< 3** (instead of3 seconds) : "SELECT book_read,title,author,publisher,pdate,place,edition,category,language,id,price,monetary_units,quantity,binding_type,location,isbn13,lccontrolnumber,callnumber,deweynumber,originality,condition,accession_number FROM book"

by sql about **<7** (instead of6 seconds) for "SELECT * FROM book" on 30 columns

once again a big bravo!

**Comment by textbrowser** [comment]
5.3 seconds.

![now](https://user-images.githubusercontent.com/10701156/220212325-60c90b23-efbb-48e3-89f8-05654383c16e.png)

**Comment by textbrowser** [comment]
With intelligent image caching. 1.5 seconds.

![now](https://user-images.githubusercontent.com/10701156/220217638-26689c2a-569e-4e33-b436-81021a70d491.png)

**Comment by textbrowser** [comment]
38466 saturn   ./BiblioteQ                        0   131.3M   140.9M   204.9M 
38756 saturn   ./BiblioteQ                        0   769.6M   779.0M   842.4M 

First entry reflects image caching. BQ consuming 131.3 MiB.
Second entry reflects the current implementation. BQ consuming 770 MiB.

Image caching is not a complete solution because unique images will require individual resources.
